cpp/src/arrow/array/builder_nested.h

```diff
@@ -45,8 +45,14 @@ class ARROW_EXPORT ListBuilder : public ArrayBuilder {
  public:
   /// Use this constructor to incrementally build the value array along with offsets and
   /// null bitmap.
+  ///
+  /// \param[in] pool the memory pool to use
+  /// \param[in] value_builder array builder for the value array
+  /// \param[in] type data type of the list that is being built
+  /// \param[in] infer_type if true, will try to infer the data type in the Finish
+  /// method. This is EXPERIMENTAL.
   ListBuilder(MemoryPool* pool, std::shared_ptr<ArrayBuilder> const& value_builder,
-              const std::shared_ptr<DataType>& type = NULLPTR);
+              const std::shared_ptr<DataType>& type = NULLPTR, bool infer_type = false);
```

Comment:
I don't understand why you are adding this? You already have `type = NULLPTR` for the inference case.

Reply:
Ok, I'm happy to remove it; this is to make sure .type() is not called when the type is nullptr (just to make it more explicit), and also to make clear that this is experimental. But really I'm fine either way.

Reply:
I'd rather have it removed indeed. It mostly seems to add complication in the API.

Reply:
SGTM